R/test-compiled-code.r

```diff
@@ -0,0 +1,99 @@
+parse_linking_to <- function(pkg_path) {
+
+  if (!file.exists(file.path(pkg_path, "DESCRIPTION")))
+    stop("No 'DESCRIPTION' file for directory '", basename(pkg_path), "'")
+
+  DESCRIPTION <- read.dcf(file.path(pkg_path, "DESCRIPTION"), all = TRUE)
+  LinkingTo <- DESCRIPTION$LinkingTo
+
+  if (is.null(LinkingTo))
+    return(character())
+
+  splat <- strsplit(LinkingTo, "\\s*[,\\n]\\s*", perl = TRUE)[[1]]
+
+  # Remove any version information
+  splat <- gsub("\\s*\\(.*", "", splat)
+
+  splat
+
+}
+
+test_compiled_code <- function(package, test_path, filter) {
+
+  pkg_path <- system.file(package = package)
+  test_path <- file.path(pkg_path, "tests", "testthat", "cpp")
+
+  # Get all of the C++ files in the directory
+  test_files <- list.files(test_path,
+                           pattern = "^test-",
+                           full.names = TRUE)
+
+  ## Generate a script that will compile an executable
+  ## to run the tests.
+
+  # Get the main.cpp wrapper
+  main <- system.file(package = "testthat", "resources", "main.cpp")
+
+  # Compile the executable
+  random_string <- paste(sample(c(0:9, letters, LETTERS), 16, replace = TRUE), collapse = "")
+  bin_path <- file.path(
+    tempdir(),
+    paste("testthat-bin-", random_string, sep = "")
+  )
+
+  if (file.exists(bin_path))
+    stop("Directory '", bin_path, "' already exists; collision with other running tests?")
+
+  dir.create(bin_path, recursive = TRUE)
+  on.exit(unlink(bin_path, recursive = TRUE), add = TRUE)
+
+  # TODO: choose appropriate compiler?
+  compiler <- if (Sys.info()[["sysname"]] == "Darwin")
```

Comment:
I went the full 9 yards and wrote a simple parser for the various `Makevars` that could be used to set this as well :)